--- a/drivers/net/phy/ar40xx.c
+++ b/drivers/net/phy/ar40xx.c
@@ -879,6 +879,55 @@ ar40xx_sw_set_port_led(struct switch_dev
 	return 0;
 }
 
+static int
+ar40xx_sw_get_port_enable(struct switch_dev *dev,
+		       const struct switch_attr *attr,
+		       struct switch_val *val)
+{
+	struct ar40xx_priv *priv = swdev_to_ar40xx(dev);
+
+	if (val->port_vlan >= AR40XX_NUM_PHYS)
+		return -EINVAL;
+
+	struct mii_bus *bus = priv->mii_bus;
+
+	int control = mdiobus_read(bus, val->port_vlan, AR40XX_CONTROL);
+
+	val->value.i = !((control & AR40XX_CONTROL_POWER_DOWN_MASK) >> AR40XX_CONTROL_POWER_DOWN_SHIFT);
+
+	return 0;
+}
+
+static int
+ar40xx_sw_set_port_enable(struct switch_dev *dev,
+		       const struct switch_attr *attr,
+		       struct switch_val *val)
+{
+	struct ar40xx_priv *priv = swdev_to_ar40xx(dev);
+
+	if (val->port_vlan >= AR40XX_NUM_PHYS)
+		return -EINVAL;
+
+	struct mii_bus *bus = priv->mii_bus;
+
+	int control = mdiobus_read(bus, val->port_vlan, AR40XX_CONTROL);
+
+	switch (val->value.i) {
+	case AR40XX_PORT_DISABLED:
+		control |= AR40XX_CONTROL_POWER_DOWN_MASK;
+		break;
+	case AR40XX_PORT_ENABLED:
+		control &= ~AR40XX_CONTROL_POWER_DOWN_MASK;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	mdiobus_write(bus, val->port_vlan, AR40XX_CONTROL, control);
+
+	return 0;
+}
+
 static void
 ar40xx_read_port_link(struct ar40xx_priv *priv, int port,
 		      struct switch_port_link *link)
@@ -1106,7 +1155,14 @@ static const struct switch_attr ar40xx_s
 		.description = "Get/Set port led mode",
 		.set = ar40xx_sw_set_port_led,
 		.get = ar40xx_sw_get_port_led,
-	}
+	},
+	{
+		.type = SWITCH_TYPE_INT,
+		.name = "enable",
+		.description = "Enable/Disable port",
+		.set = ar40xx_sw_set_port_enable,
+		.get = ar40xx_sw_get_port_enable,
+	},
 };
 
 const struct switch_attr ar40xx_sw_attr_vlan[] = {
--- a/drivers/net/phy/ar40xx.h
+++ b/drivers/net/phy/ar40xx.h
@@ -297,6 +297,10 @@ struct ar40xx_mib_desc {
 #define PHY4_PREFER_COPPER		0x0
 #define PHY4_PREFER_FIBER		0x1
 
+#define AR40XX_CONTROL 0x0
+#define   AR40XX_CONTROL_POWER_DOWN_SHIFT	11
+#define   AR40XX_CONTROL_POWER_DOWN_MASK	BIT(AR40XX_CONTROL_POWER_DOWN_SHIFT)
+
 /* port forwarding state */
 enum {
 	AR40XX_PORT_STATE_DISABLED = 0,
@@ -336,6 +340,12 @@ enum {
 	AR40XX_PORT_LED_MODE_STATUS = 2
 };
 
+/* port state */
+enum {
+	AR40XX_PORT_DISABLED = 0,
+	AR40XX_PORT_ENABLED = 1
+};
+
 #define AR40XX_MIB_WORK_DELAY	2000 /* msecs */
 
 #define AR40XX_QM_WORK_DELAY    100
