--- a/state.c
+++ b/state.c
@@ -97,6 +97,8 @@ static void set_console(void)
 
 static void perform_halt()
 {
+	int cnt = 0;
+
 	if (reboot_event == RB_POWER_OFF)
 		LOG("- power down -\n");
 	else
@@ -105,22 +107,39 @@ static void perform_halt()
 	/* Allow time for last message to reach serial console, etc */
 	sleep(1);
 
+	LOG("- pre-container -\n");
+
 	if (is_container()) {
 		reboot(reboot_event);
 		exit(EXIT_SUCCESS);
 		return;
 	}
 
+	LOG("- pre-vfork -\n");
+
 	/* We have to fork here, since the kernel calls do_exit(EXIT_SUCCESS)
 	 * in linux/kernel/sys.c, which can cause the machine to panic when
 	 * the init process exits... */
 	if (!vfork()) { /* child */
+		LOG("- pre-reboot -\n");
 		reboot(reboot_event);
 		_exit(EXIT_SUCCESS);
 	}
 
-	while (1)
+	LOG("- post vfork -\n");
+
+	while (1) {
+		if (cnt == 5) {
+			LOG("VFORK FAILED, DOING REBOOT()\n");
+			sleep(1);
+			cnt = 0;
+
+			reboot(reboot_event);
+			LOG("RETRYING...\n");
+		}
+		cnt++;
 		sleep(1);
+	}
 }
 
 static void state_enter(void)
